### PR TITLE
fix: Store Cert file in the charm container and provide the path to tempo

### LINKED ISSUE #208: missing argument to charm_tracing breaks the charm when integrated to Tempo
### Bug Description

the trace_charm decorator will raise an exception if you pass to it an https `endpoint` but no `server_cert`.

### To Reproduce

juju deploy self-signed-certificates ssc
juju deploy traefik-k8s traefik  # might not be necessary
juju deploy tempo-k8s tempo

jhack imatrix fill

### Environment

juju 3.5

### Relevant log output

```shell
unit-ssc-0: 11:55:39 ERROR unit.ssc/0.juju-log certificates:39: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-ssc-0/charm/./src/charm.py", line 341, in <module>
    main(SelfSignedCertificatesCharm)
  File "/var/lib/juju/agents/unit-ssc-0/charm/venv/ops/main.py", line 549, in main
    manager = _Manager(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-ssc-0/charm/venv/ops/main.py", line 432, in __init__
    self.charm = self._make_charm(self.framework, self.dispatcher)
  File "/var/lib/juju/agents/unit-ssc-0/charm/venv/ops/main.py", line 435, in _make_charm
    charm = self._charm_class(framework)
  File "/var/lib/juju/agents/unit-ssc-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 450, in wrap_init
    raise TLSError(
charms.tempo_k8s.v1.charm_tracing.TLSError: Tracing endpoint is https, but no server_cert has been passed.Please point @trace_charm to a `s
rver_cert` attr.
unit-ssc-0: 11:55:39 ERROR juju.worker.uniter.operation hook "certificates-relation-changed" (via hook dispatching script: dispatch) failed
 exit status 1
unit-ssc-0: 11:55:39 INFO juju.worker.uniter awaiting error resolution for "relation-changed" hook
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

Store Cert file in the charm workload and provide the path to tempo

Fixes #208 

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -174,6 +183,7 @@ def _generate_root_certificate(self) -> None:
             secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
             secret.set_content(content=secret_content)
         else:
+            self._push_ca_cert_to_container(ca_certificate.decode())
```
Comment (DanielArndt) [inline src/charm.py:186]: Shouldn't we do this in both cases (outside the if)?  Otherwise we won't update it on the workload if the certificate changes.

**Review by DanielArndt — COMMENTED** [review]
LGTM from a code perspective (except this one comment which I'm pretty sure is incorrect?).  But, I'm not sure I grasp the consequences of this change.

I've not really done any work with workload-less charms, so I'm fairly ignorant. Does it matter that we now have a workload, on a charm that was previously a workload-less charm? Although, it seems all we're doing is using it for storage...

File: src/charm.py (hunk 1 above)
Comment (saltiyazan) [reply src/charm.py]: Yes, good point. Fixed.

**Comment by saltiyazan** [comment]
> LGTM from a code perspective (except this one comment which I'm pretty sure is incorrect?). But, I'm not sure I grasp the consequences of this change.
> 
> I've not really done any work with workload-less charms, so I'm fairly ignorant. Does it matter that we now have a workload, on a charm that was previously a workload-less charm? Although, it seems all we're doing is using it for storage...

We don't add a workload we just attach storage and store the CA cert file there in the charm container (fixed in the subject of the PR too). It would be simpler if we could pass the CA certificate itself to tempo instead of file path. Looking at the lib I don't think it is possible? is that right @PietroPasotti ?

**Review by DanielArndt — APPROVED** [review]

**Comment by PietroPasotti** [comment]
Hi @saltiyazan! Thanks for this and sorry for the delay in the review.
I think this approach is fine, I'm wondering if you really need to add a storage only for this one file? Can't the charm write it to /tmp/ and be done with it? After all it's only used by charm_tracing.

As to passing the cert itself (instead of a path to a file) to charm_tracing, that's not possible since the instrumentation library we use (the otlp sdk) requires a file path and we'd rather have the charm control where it's stored.

**Comment by saltiyazan** [comment]
> Hi @saltiyazan! Thanks for this and sorry for the delay in the review. I think this approach is fine, I'm wondering if you really need to add a storage only for this one file? Can't the charm write it to /tmp/ and be done with it? After all it's only used by charm_tracing.
> 
> As to passing the cert itself (instead of a path to a file) to charm_tracing, that's not possible since the instrumentation library we use (the otlp sdk) requires a file path and we'd rather have the charm control where it's stored.

Thanks @PietroPasotti. That probably is a better approach.

**Review by kayra1 — APPROVED** [review]
much nicer this way even if it's a bit unusual

**Pull request merged**